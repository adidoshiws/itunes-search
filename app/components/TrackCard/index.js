/**
 *
 * Track Card
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card } from 'antd';
import T from '@components/T';
import If from '@components/If';
import { isEmpty } from 'lodash';

const CustomCard = styled(Card)`
  && {
    margin: 20px 0;
  }
`;

export function TrackCard({ trackName, collectionName, previewUrl }) {
  return (
    <CustomCard data-testid="track-card">
      <If condition={!isEmpty(trackName)} otherwise={<T data-testid="name-unavailable" id="track_name_unavailable" />}>
        <T data-testid="name" id="track_name" values={{ trackName: trackName }} />
      </If>
      <If
        condition={!isEmpty(collectionName)}
        otherwise={<T data-testid="collectionName-unavailable" id="track_collection_name_unavailable" />}
      >
        <T data-testid="collectionName" id="track_collection_name" values={{ collectionName: collectionName }} />
      </If>
      <If condition={previewUrl} otherwise={<T data-testid="preview-unavailable" id="track_preview_unavailable" />}>
        <T data-testid="preview" id="track_preview" values={{ preview: previewUrl }} />
      </If>
    </CustomCard>
  );
}

TrackCard.propTypes = {
  trackName: PropTypes.string,
  collectionName: PropTypes.string,
  previewUrl: PropTypes.string
};

export default TrackCard;
