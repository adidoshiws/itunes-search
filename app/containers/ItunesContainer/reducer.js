import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: itunesContainerTypes, Creators: itunesContainerCreators } = createActions({
  requestGetItuneTracks: ['trackName'],
  successGetItuneTracks: ['data'],
  failureGetItuneTracks: ['error'],
  clearItuneTracks: {}
});
export const initialState = { trackName: null, tracksData: {}, tracksError: null };

export const itunesContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case itunesContainerTypes.REQUEST_GET_ITUNE_TRACKS:
        draft.trackName = action.trackName;
        break;
      case itunesContainerTypes.CLEAR_ITUNE_TRACKS:
        draft.trackName = null;
        draft.tracksError = null;
        draft.tracksData = {};
        break;
      case itunesContainerTypes.SUCCESS_GET_ITUNE_TRACKS:
        draft.tracksData = action.data;
        break;
      case itunesContainerTypes.FAILURE_GET_ITUNE_TRACKS:
        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default itunesContainerTypes;
