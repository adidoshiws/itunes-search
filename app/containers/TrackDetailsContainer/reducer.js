import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackDetailsTypes, Creators: trackDetailsCreators } = createActions({
  requestGetTrackDetails: ['trackId'],
  successGetTrackDetails: ['data'],
  failureGetTrackDetails: ['error']
});
export const initialState = {
  trackId: null,
  trackDetailsData: {},
  trackDetailsError: null
};

export const trackDetailsReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackDetailsTypes.REQUEST_GET_TRACK_DETAILS:
        draft.trackId = action.trackId;
        break;
      case trackDetailsTypes.SUCCESS_GET_TRACK_DETAILS:
        draft.trackDetailsData = action.data;
        break;
      case trackDetailsTypes.FAILURE_GET_TRACK_DETAILS:
        draft.trackDetailsError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default trackDetailsTypes;
