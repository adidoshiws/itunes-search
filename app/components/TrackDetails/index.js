/**
 *
 * Track Details
 *
 */
import React, { useEffect, useState, useRef } from 'react';
import { useParams } from 'react-router-dom';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { selectTrackDetailsData, selectTrackDetailsError } from '@app/containers/ItunesContainer/selectors';
import { itunesContainerCreators } from '@app/containers/ItunesContainer/reducer';
import itunesContainerSaga from '@app/containers/ItunesContainer/saga';
import { injectSaga } from 'redux-injectors';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Image, Button } from 'antd';
import T from '@components/T';
import If from '@components/If';
import isEmpty from 'lodash/isEmpty';
import { injectIntl, FormattedMessage } from 'react-intl';

const Container = styled.div`
  && {
    margin: 1.9rem auto;
    padding: 20px;
    height: max-content;
    width: 40rem;
    box-shadow: 0 5px 20px -5px rgb(191 192 192);
    border-radius: 0.8rem;
  }
`;

const ImageWrapper = styled.div`
  display: flex;
  gap: 1.4rem;
`;

const TrackDetailsSection = styled.div``;

const TrackImage = styled(Image)`
  width: ${(props) => props.width}px;
`;

const PlayBtn = styled(Button)`
  type: ${(props) => props.type};
`;

export function TrackDetails({ trackDetailsData, dispatchItuneTrackDetails }) {
  const [loading, setLoading] = useState(false);
  const [isPlaying, setIsPlaying] = useState(false);
  const params = useParams();
  const audioRef = useRef();

  const { results } = trackDetailsData;

  const playHandler = () => {
    if (audioRef.current.paused) {
      audioRef.current.play();
      setIsPlaying(true);
    } else {
      audioRef.current.pause();
      setIsPlaying(false);
    }
  };

  useEffect(() => {
    if (trackDetailsData) {
      setLoading(false);
    }
  }, [trackDetailsData]);

  useEffect(() => {
    if (params.trackId) {
      setLoading(true);
      dispatchItuneTrackDetails(params.trackId);
    }
  }, [params.trackId]);

  useEffect(() => {
    if (!trackDetailsData?.results?.length) {
      setLoading(true);
    }
  }, []);

  const details = results[0];

  const renderDetails = () => {
    <If condition={!isEmpty(details) || loading}>Loading...</If>;
  };

  return (
    <>
      <Container>
        <ImageWrapper>
          <TrackImage width={200} src={details.artworkUrl100} />
          <TrackDetailsSection>
            <If
              condition={!isEmpty(details.trackName)}
              otherwise={<T data-testid="name-unavailable" id="track_name_unavailable" />}
            >
              <T data-testid="trackName" id="track_name" values={{ trackName: details.trackName }} />
            </If>

            <If
              condition={!isEmpty(details.collectionName)}
              otherwise={<T data-testid="collectionName-unavailable" id="track_collection_name_unavailable" />}
            >
              <T
                data-testid="collectionName"
                id="track_collection_name"
                values={{ collectionName: details.collectionName }}
              />
            </If>

            <If
              condition={!isEmpty(details.country)}
              otherwise={<T data-testid="country-unavailable" id="track_country_unavailable" />}
            >
              <T data-testid="trackCountry" id="track_country" values={{ country: details.country }} />
            </If>

            <If
              condition={!isEmpty(details.kind)}
              otherwise={<T data-testid="kind-unavailable" id="track_kind_unavailable" />}
            >
              <T data-testid="trackKind" id="track_kind" values={{ kind: details.kind }} />
            </If>
            <audio ref={audioRef} data-testid="preview" id="track_preview" src={details.previewUrl}></audio>
            <PlayBtn type="primary" onClick={playHandler}>
              {isPlaying ? <FormattedMessage id="details_pause_btn" /> : <FormattedMessage id="details_play_btn" />}
            </PlayBtn>
          </TrackDetailsSection>
        </ImageWrapper>
      </Container>
      {renderDetails()}
    </>
  );
}

TrackDetails.propTypes = {
  trackDetailsData: PropTypes.object,
  dispatchItuneTrackDetails: PropTypes.func
};

const mapStateToProps = createStructuredSelector({
  trackDetailsData: selectTrackDetailsData(),
  trackDetailsError: selectTrackDetailsError()
});

export function mapDispatchToProps(dispatch) {
  const { getTrackDetails } = itunesContainerCreators;
  return {
    dispatchItuneTrackDetails: (trackId) => dispatch(getTrackDetails(trackId))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  injectIntl,
  withConnect,
  injectSaga({ key: 'itunesContainer', saga: itunesContainerSaga })
)(TrackDetails);

export const TrackDetailsTest = compose(injectIntl)(TrackDetails);
