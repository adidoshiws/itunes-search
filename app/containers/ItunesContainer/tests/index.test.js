/**
 *
 * Tests for ItunesContainer
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { timeout, renderProvider } from '@utils/testUtils';
import { ItunesContainerTest as ItunesContainer, mapDispatchToProps } from '../index';
import { itunesContainerTypes } from '../reducer';
import { translate } from '@app/components/IntlGlobalProvider';

describe('<ItunesContainer /> tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ItunesContainer dispatchItuneTracks={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearItuneTracks on empty change', async () => {
    const getItuneTracksSpy = jest.fn();
    const clearItuneTracksSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <ItunesContainer dispatchClearItuneTracks={clearItuneTracksSpy} dispatchItuneTracks={getItuneTracksSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'a' }
    });
    await timeout(500);
    expect(getItuneTracksSpy).toBeCalled();
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearItuneTracksSpy).toBeCalled();
  });

  it('should call dispatchItuneTracks on change and after enter', async () => {
    const trackName = 'rockstar';
    const { getByTestId } = renderProvider(<ItunesContainer dispatchItuneTracks={submitSpy} />);
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: { value: trackName }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith(trackName);

    fireEvent.keyDown(searchBar, {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(submitSpy).toBeCalledWith(trackName);
  });

  it('should  dispatchItuneTracks on update on mount if trackName is already persisted', async () => {
    const trackName = 'rockstar';
    renderProvider(<ItunesContainer trackName={trackName} tracksData={null} dispatchItuneTracks={submitSpy} />);

    await timeout(500);
    expect(submitSpy).toBeCalledWith(trackName);
  });

  // check
  it('should validate mapDispatchToProps actions', async () => {
    const dispatchTracksSearchSpy = jest.fn();
    const trackName = 'rockstar';
    const actions = {
      dispatchItuneTracks: { trackName, type: itunesContainerTypes.REQUEST_GET_ITUNE_TRACKS },
      dispatchClearItuneTracks: { type: itunesContainerTypes.CLEAR_ITUNE_TRACKS }
    };

    const props = mapDispatchToProps(dispatchTracksSearchSpy);
    props.dispatchItuneTracks(trackName);
    expect(dispatchTracksSearchSpy).toHaveBeenCalledWith(actions.dispatchItuneTracks);

    await timeout(500);
    props.dispatchClearItuneTracks();
    expect(dispatchTracksSearchSpy).toHaveBeenCalledWith(actions.dispatchClearItuneTracks);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(<ItunesContainer tracksError={defaultError} />);
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });

  it('should render the default message when searchBox is empty and tracksError is null', () => {
    const defaultMessage = translate('track_search_default');
    const { getByTestId } = renderProvider(<ItunesContainer />);
    expect(getByTestId('default-message')).toBeInTheDocument();
    expect(getByTestId('default-message').textContent).toBe(defaultMessage);
  });

  it('should render the data when loading becomes false', () => {
    const tracksData = { results: [{ trackOne: 'rockstar' }] };
    const { getByTestId } = renderProvider(<ItunesContainer tracksData={tracksData} dispatchItuneTracks={submitSpy} />);
    expect(getByTestId('for')).toBeInTheDocument();
  });

  it('should render exact number of TrackCards as per resultCount in results', () => {
    const resultCount = 3;
    const tracksData = {
      resultCount,
      results: [
        {
          trackName: 'broken',
          collectionName: 'Jack Johnson and Friends',
          preview:
            'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/8b/fd/81/8bfd81a6-0217-b56c-01b0-7249ed910a35/mzaf_16218415284344588014.plus.aac.p.m4a'
        },
        {
          trackName: 'react',
          collectionName: 'Jack Johnson and Friends',
          preview:
            'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/0d/1b/1a/0d1b1a37-2068-0e62-2162-0c9822fbc03c/mzaf_4390099184106020569.plus.aac.p.m4a'
        },
        {
          trackName: 'my own two hands',
          collectionName: 'Three Way Tie',
          preview:
            'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/fd/bb/38/fdbb38d2-073d-4bc7-68c4-348a0be6d560/mzaf_4150435585996894188.plus.aac.p.m4a'
        }
      ]
    };
    const { getAllByTestId } = renderProvider(
      <ItunesContainer tracksData={tracksData} dispatchItuneTracks={submitSpy} />
    );
    expect(getAllByTestId('track-card').length).toBe(resultCount);
  });

  it('should render Skeleton Comp when "loading" is true', async () => {
    const trackName = 'some track';
    const { getByTestId, baseElement } = renderProvider(
      <ItunesContainer dispatchItuneTracks={submitSpy} trackName={trackName} />
    );
    fireEvent.change(getByTestId('search-bar'), { target: { value: trackName } });
    await timeout(500);
    expect(baseElement.getElementsByClassName('ant-skeleton').length).toBe(1);
  });
});
