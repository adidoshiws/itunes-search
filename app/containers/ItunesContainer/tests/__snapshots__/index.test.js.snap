// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItunesContainer /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <input
      data-testid="search-bar"
      placeholder="search tracks"
      type="text"
      value=""
    />
  </div>
</body>
`;
